nmagent/internal/validate.go

```diff
@@ -0,0 +1,55 @@
+package internal
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// ValidationError {{{1
+
+type ValidationError struct {
+	MissingFields []string
+}
+
+func (v ValidationError) Error() string {
+	return fmt.Sprintf("missing fields: %s", strings.Join(v.MissingFields, ", "))
+}
+
+func (v ValidationError) IsEmpty() bool {
+	return len(v.MissingFields) == 0
+}
+
+// }}}1
+
+// Validate searches for validate struct tags and performs the validations
+// requested by them
+func Validate(obj interface{}) error {
+	errs := ValidationError{}
+
+	val := reflect.ValueOf(obj)
+	typ := reflect.TypeOf(obj)
+
+	for i := 0; i < val.NumField(); i++ {
+		fieldVal := val.Field(i)
+		fieldTyp := typ.Field(i)
+
+		op := fieldTyp.Tag.Get("validate")
+		switch op {
+		case "presence":
```

Comment:
i'm not convinced this is necessary. we are the consumers of this package and know what inputs it needs

Reply:
So this came about from porting the existing validations over. I had a bunch of logic in the `Validate` methods for each request that looked something like:

```go
if r.SomeField == "" {
       err.MissingFields = append(err.MissingFields, "SomeField")
}

// repeat for a bunch of other stuff
```

`Validate` is another method, so when it came time to answer the question "what fields are actually required for this request?" I had to go digging into that method to figure it out. I started adding comments to the struct saying like "this field is required"... which led me to think "well, a struct tag can serve this purpose too, and it can actually do the validation." I initially thought `required:"true"`, but that brought about the question "what does `required:"false"` mean and how is that distinct from not providing it at all?" (it isn't). So I settled on `validate:"presence"`, since it does allow pulling in other common validation as well. The `Validate` methods still exist as a place for anything particularly weird, but they're pretty vestigial at this point.

It's kinda fancy, true, but it was born out of a practical refactoring. Plus I can snag the name of the field out with reflection rather than having to remember to update the string if it ever gets renamed.